pkg/app/ops/pipedstatsbuilder/builder.go

```diff
@@ -0,0 +1,57 @@
+// Copyright 2021 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pipedstatsbuilder
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"io/ioutil"
+
+	"go.uber.org/zap"
+
+	"github.com/pipe-cd/pipe/pkg/cache"
+)
+
+type PipedStatsBuilder struct {
+	backend cache.Cache
+	logger  *zap.Logger
+}
+
+func NewPipedStatsBuilder(c cache.Cache, logger *zap.Logger) *PipedStatsBuilder {
+	return &PipedStatsBuilder{
+		backend: c,
+		logger:  logger.Named("piped-metrics-builder"),
+	}
+}
+
+func (b *PipedStatsBuilder) Build() (io.ReadCloser, error) {
+	res, err := b.backend.GetAll()
+	if err != nil {
+		b.logger.Error("failed to fetch piped stats from cache", zap.Error(err))
+		return nil, err
+	}
+	data := make([][]byte, 0, len(res))
+	for _, v := range res {
+		value, okValue := v.([]byte)
+		if !okValue {
+			err = errors.New("error value not a bulk of string value")
+			b.logger.Error("failed to marshal piped stat data", zap.Error(err))
+			return nil, err
+		}
+		data = append(data, value)
+	}
+	return ioutil.NopCloser(bytes.NewReader(bytes.Join(data, []byte("\n")))), nil
```

Comment:
iouti.NopClose is deprecated. Instead using io.NopCloser is preferable.
https://golang.org/pkg/io/ioutil/#NopCloser

Reply:
Well, why does it need to be an io.ReadCloser? Seems like io.Reader is enough.

Reply:
👀  Just want to gracefully close the []byte stream on the hander side. Or is it unnecessary? 🤔 
ref: https://github.com/pipe-cd/pipe/pull/2202/files#diff-c9b9fc9a691a60f03cc1d752ce9fac11e932d6a2867c984bd26923131704fdeaR152

Reply:
First, NopCloser.Close does nothing.

And with the current implementation of `PipedStatsBuilder.Build()`, all byte sequences are on the heap. So no need to do close operations at all.

Reply:
If you're planning to directly read from any kinds of stream like Redis, then you will have to add the close operation.

Reply:
I see 👍  Since we will keep the GetAll interface which returns data ( not the reader or else to direct stream data ), it's safe to remove this unnecessary closer as you mentioned 🙏  Lets me address it here 🙆‍♀️

Reply:
Addressed on https://github.com/pipe-cd/pipe/pull/2202/commits/d4489194fe41c2bb1f46967f251b818c8b924833 🙏